alg/viewshed.h

```diff
@@ -0,0 +1,96 @@
+/******************************************************************************
+ *
+ * Project:  Viewshed Generation
+ * Purpose:  Core algorithm implementation for viewshed generation.
+ * Author:   Tamas Szekeres, szekerest@gmail.com
+ *
+ ******************************************************************************
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include "cpl_progress.h"
+#include "gdal_priv.h"
+
+namespace gdal
+{
+
+class Viewshed
```

Comment:
Doxygen isn't happy about of the lack of documentation: https://github.com/OSGeo/gdal/actions/runs/8909043956/job/24465719151?pr=9827

For non-public API, we don't necessarily require full Doxygen docs. You may mark the file as:
```
//! @cond Doxygen_Suppress
[ ... ]
//! @endcond
```

you may also need to surround methods in the .cpp with that, but I never know exactly when this is needed.

You may run locally "make doxygen" in doc/ to check if it is happy.
The CI check actually does:
```
make .doxygen_up_to_date >log.txt 2>&1
if grep -i warning log.txt | grep -v -e russian -e brazilian; then echo "Doxygen warnings found" && cat log.txt && /bin/false; else echo "No Doxygen warnings found"; fi
```